File: tools/EventClients/examples/python/example_button2.py
#!/usr/bin/python

# This is a simple example showing how you can send a key press event
# to XBMC in a non-queued fashion to achieve a button pressed down
# event i.e. a key press that repeats.

# The repeat interval is currently hard coded in XBMC but that might
# change in the future.

# NOTE: Read the comments in 'example_button1.py' for a more detailed
# explanation.

import sys
sys.path.append("../../lib/python")

from xbmcclient import *
from socket import *

def main():
    import time
    import sys

    host = "localhost"
    port = 9777
    addr = (host, port)

    sock = socket(AF_INET,SOCK_DGRAM)

    # First packet must be HELO and can contain an icon
    packet = PacketHELO("Example Remote", ICON_PNG,
                        "../../icons/bluetooth.png")
    packet.send(sock, addr)

    # wait for notification window to close (in XBMC)
    time.sleep(5)

    # send a up key press using the xbox gamepad map "XG" and button
    # name "dpadup" ( see PacketBUTTON doc for more details)
    packet = PacketBUTTON(map_name="XG", button_name="dpadup")
    packet.send(sock, addr)

    # wait for a few seconds to see its effect
    time.sleep(5)

    # send a down key press using the raw keyboard code
    packet = PacketBUTTON(code=0x28)
    packet.send(sock, addr)

    # wait for a few seconds to see its effect
    time.sleep(5)

    # send a right key press using the keyboard map "KB" and button
    # name "right"
    packet = PacketBUTTON(map_name="KB", button_name="right")
    packet.send(sock, addr)

    # wait for a few seconds to see its effect
    time.sleep(5)

    # that's enough, release the button. During release, button code
    # doesn't matter.
    packet = PacketBUTTON(code=0x28, down=0)
    packet.send(sock, addr)

    # ok we're done, close the connection
    # Note that closing the connection clears any repeat key that is
    # active. So in this example, the actual release button event above
    # need not have been sent.
    packet = PacketBYE()
    packet.send(sock, addr)

if __name__=="__main__":
    main()

File: tools/EventClients/examples/c++/example_button2.cpp
#include "../../lib/c++/xbmcclient.h"
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#ifdef _WIN32
#include <Windows.h> // for sleep
#else
#include <unistd.h>
#endif

int main(int argc, char **argv)
{
  /* connect to localhost, port 9777 using a UDP socket
     this only needs to be done once.
     by default this is where XBMC will be listening for incoming
     connections. */
  CAddress my_addr; // Address => localhost on 9777
  int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
  if (sockfd < 0)
  {
    printf("Error creating socket\n");
    return -1;
  }

  my_addr.Bind(sockfd);

  std::string sIconFile = "../../icons/bluetooth.png";
  unsigned short usIconType = ICON_PNG;

  std::ifstream file (sIconFile, std::ios::in|std::ios::binary|std::ios::ate);
  if (!file.is_open())
  {
    sIconFile = "/usr/share/pixmaps/kodi/bluetooth.png";
    file.open(sIconFile, std::ios::in|std::ios::binary|std::ios::ate);

    if (!file.is_open()) {
      usIconType = ICON_NONE;
    }
    else
    {
      file.close();
    }
  }
  else
  {
    file.close();
  }

  CPacketHELO HeloPackage("Example Remote", usIconType, sIconFile.c_str());
  HeloPackage.Send(sockfd, my_addr);

  sleep(5);
  // Note that we have foo(BUTTON, DEVICEMAP);
  CPacketBUTTON btn1("dpadup", "XG", 0);
  btn1.Send(sockfd, my_addr);

  sleep(5);

  CPacketBUTTON btn2(0x28, 0);
  btn2.Send(sockfd, my_addr);

  sleep(5);

  CPacketBUTTON btn3("right", "KB", 0);
  btn3.Send(sockfd, my_addr);

  sleep(5);
  // Release button
  CPacketBUTTON btn4;
  btn4.Send(sockfd, my_addr);

  // BYE is not required since XBMC would have shut down
  CPacketBYE bye; // CPacketPing if you want to ping
  bye.Send(sockfd, my_addr);
}
